### ISSUE TITLE
Regarding consensus posterior calculation

### ISSUE THREAD

**Issue by zedoul** [strongh/stark#1]
Hi,

In stark.py, line number 50, we have a consensus posterior calculation as follows:

```
return np.dot(w_j[0], f1) + np.dot(w_j[1], f2)
```

And according to the papar (http://research.google.com/pubs/pub41849.html), it should be calculated by (3) in page 6

$\theta_g = (\sum_s W_s)^{-1} \sum_s W_s \theta_{s,g}$ (3)

The problem is that they are not equal to each other. So I guess we need to fix the script as follows:

from     

```
 return np.dot(w_j[0], f1) + np.dot(w_j[1], f2)
```

to     

```
 l = np.linarg.inv(w_j[0] + w_j[1])
 r = np.dot(w_j[0], f1) + np.dot(w_j[1], f2)
 return np.dot(l, r)
```

I hope to have your feedback for this suggestion.

**Comment by strongh** [comment]
Sorry for the delay in addressing your question, @zedoul!

I certainly agree that the calculation you mentioned was incorrect. However I don't think your suggestion would have correctly handled the calculations. The problem is that that weighting calculation is performed in a pairwise fashion (as a Spark reduce), whereas the formula assumes that the inverse is of the sum over all the weights. Essentially the problem is that the inverse of the sum is not the sum of the inverses. This was already a problem in the initial implementation.

I've just committed a possible fix, in eb7ffedd1d87d7ab6c863ba45d53be5eae92f1b0, which follows the paper you mentioned. It's actually simpler than before because Scott's version ignores the influence of the prior. The way I got around the issue above is that the reduce step calculates two matrices instead of one. Rather than directly returning the consensus posterior as before, the new reducer returns an `l` and an `r`, similar to yours (except that my reduce `l` does not been inverted). Then the final inversion and matrix product is calculated by the driver program. This should still be fairly scalable since those two matrices will be the same size regardless of how many partitions there are.

Thanks for your interest! I hope this latest change makes sense to you.
